ocaml/runtime/float32.c

```diff
@@ -0,0 +1,158 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*                      Max Slater, Jane Street                           */
+/*                                                                        */
+/*   Copyright 2024 Jane Street Group LLC                                 */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#include <math.h>
+#include <float.h>
+
+#include "caml/alloc.h"
+#include "caml/custom.h"
+#include "caml/float32.h"
+#include "caml/memory.h"
+#include "caml/intext.h"
+#include "caml/mlvalues.h"
+
+#define CAML_INTERNALS
+
+CAML_STATIC_ASSERT(sizeof(float) == sizeof(int32_t));
+
+static int32_t float32_as_int32(float f)
+{
+  union { float f; int32_t i; } u;
+  u.f = f;
+  return u.i;
+}
+
+intnat caml_float32_compare_unboxed(float f, float g)
+{
+  /* If one or both of f and g is NaN, order according to the convention
+     NaN = NaN and NaN < x for all other floats x. */
+  /* This branchless implementation is from GPR#164.
+     Note that [f == f] if and only if f is not NaN.
+     We expand each subresult of the expression to
+     avoid sign-extension on 64bit. GPR#2250.  */
+  intnat res =
+      (intnat)(f > g) - (intnat)(f < g) + (intnat)(f == f) - (intnat)(g == g);
+  return res;
+}
+
+static int float32_cmp(value v1, value v2)
+{
+  return caml_float32_compare_unboxed(Float32_val(v1), Float32_val(v2));
+}
+
+static intnat float32_hash(value v)
+{
+  float f = Float32_val(v);
+  return float32_as_int32(f);
```

Comment:
That does not feel right; for instance,
it means that `-0.0` and `+0.0` will
have different hash values.

Reply:
Updated to be consistent with `caml_hash_mix_float`